src/Storages/StorageDictionary.cpp

```diff
@@ -221,10 +223,28 @@ void StorageDictionary::renameInMemory(const StorageID & new_table_id)
             has_configuration = true;
             configuration->setString("dictionary.database", new_table_id.database_name);
             configuration->setString("dictionary.name", new_table_id.table_name);
+            if (move_between_db_engines)
+            {
+                if (new_table_id.uuid == UUIDHelpers::Nil)
+                    configuration->remove("dictionary.uuid");
+                else
+                    configuration->setString("dictionary.uuid", toString(new_table_id.uuid));
+            }
         }
     }
 
-    if (has_configuration)
+    if (!has_configuration)
+        return;
+
+    if (move_between_db_engines)
+    {
+        /// It's too hard to update both name and uuid, better to reload dictionary with new name
+        removeDictionaryConfigurationFromRepository();
+        auto repository = std::make_unique<ExternalLoaderDictionaryStorageConfigRepository>(*this);
+        remove_repository_callback = getContext()->getExternalDictionariesLoader().addConfigRepository(std::move(repository));
```

Comment:
Oh, it's a `scope_guard`... so we execute `remove` from old repository in `removeDictionaryConfigurationFromRepository` and after assign new scope_guard for new repository.